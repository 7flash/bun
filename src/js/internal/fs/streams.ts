// fs.ReadStream and fs.WriteStream are lazily loaded to avoid importing 'node:stream' until required
const { Readable, Writable } = require("node:stream");
const fs: typeof import("node:fs") = require("node:fs");
const {
  read,
  write,
  fsync,
} = fs;
const {
  FileHandle,
  kRef,
  kUnref,
  kFd,
} = (fs.promises as any).$data as {
  FileHandle: { new(): FileHandle };
  readonly kRef: unique symbol;
  readonly kUnref: unique symbol;
  readonly kFd: unique symbol;
  fs: typeof fs;
};
type FileHandle = import('node:fs/promises').FileHandle & {
  on(event: any, listener: any): FileHandle;
};
type FSStream = import("node:fs").ReadStream & import("node:fs").WriteStream & {
  fd: number | null;
  path: string;
  flags: string;
  mode: number;
  start: number;
  end: number;
  pos: number | undefined;
  bytesRead: number;
  flush: boolean;
  open: () => void;
};
type FD = number;

const { validateInteger, validateInt32, validateFunction } = require("internal/validators");

// Bun supports a fast path for `createReadStream("path.txt")` in `Bun.serve`,
// where the entire stream implementation can be bypassed, effectively making it
// `new Response(Bun.file("path.txt"))`. This makes an idomatic Node.js pattern
// much faster.
const kReadStreamFastPath = Symbol("kReadStreamFastPath");
const kFs = Symbol("kFs");
// const readStreamSymbol = Symbol.for("Bun.NodeReadStream");
// const readStreamPathOrFdSymbol = Symbol.for("Bun.NodeReadStreamPathOrFd");
// const writeStreamSymbol = Symbol.for("Bun.NodeWriteStream");
// const writeStreamPathFastPathSymbol = Symbol.for("Bun.NodeWriteStreamFastPath");
// const writeStreamPathFastPathCallSymbol = Symbol.for("Bun.NodeWriteStreamFastPathCall");
const kIoDone = Symbol("kIoDone");
const kIsPerformingIO = Symbol("kIsPerformingIO");

const { read: fileHandlePrototypeRead, write: fileHandlePrototypeWrite, fsync: fileHandlePrototypeFsync } = FileHandle.prototype;

const blobToStreamWithOffset = $newZigFunction("blob.zig", "Blob.toStreamWithOffset", 1);

const fileHandleStreamFs = (fh: FileHandle) => ({
  // try to use the basic fs.read/write/fsync if available, since they are less
  // abstractions. however, node.js allows patching the file handle, so this has
  // to be checked for.
  read: fh.read === fileHandlePrototypeRead ? read : function(fd, buf, offset, length, pos, cb) {
    return fh.read(buf,offset,length,pos).then(({ bytesRead, buffer }) => cb(null, bytesRead, buffer), (err) => cb(err, 0, buf));
  },
  write: fh.write === fileHandlePrototypeWrite ? write : function(fd, buffer, offset, length, position, cb) {
    return fh.write(buffer, offset, length, position).then(({ bytesWritten, buffer }) => cb(null, bytesWritten, buffer), (err) => cb(err, 0, buffer));
  },
  fsync: fh.sync === fileHandlePrototypeFsync ? fsync : function(fd, cb) {
    return fh.sync().then(() => cb(), cb);
  },
  close: streamFileHandleClose.bind(fh),
});

function streamFileHandleClose(this: FileHandle, fd: FD, cb: (err?: any) => void) {
  $assert(this[kFd] == fd, 'fd mismatch');
  this[kUnref]();
  this.close().then(() => cb(), cb);
}

function getValidatedPath(p: any) {
  if (p instanceof URL) return Bun.fileURLToPath(p as URL);
  if (typeof p !== "string") throw $ERR_INVALID_ARG_TYPE("path", "string or URL", p);
  return require("node:path").resolve(p);
}

function copyObject(source) {
  const target = {};
  // Node tests for prototype lookups, so { ...source } will not work.
  for (const key in source)
    target[key] = source[key];
  return target;
}

function getStreamOptions(options, defaultOptions = {}) {
  if (options == null || typeof options === 'function') {
    return defaultOptions;
  }

  if (typeof options === 'string') {
    if (options !== 'buffer' && !Buffer.isEncoding(options)) {
      throw $ERR_INVALID_ARG_VALUE("encoding", options, "is invalid encoding");
    }
    return { encoding: options };
  } else if (typeof options !== 'object') {
    throw $ERR_INVALID_ARG_TYPE('options', ['string', 'Object'], options);
  }

  let { encoding, signal = true } = options;
  if (encoding && encoding !== 'buffer' && !Buffer.isEncoding(encoding)) {
    throw $ERR_INVALID_ARG_VALUE("encoding", encoding, "is invalid encoding");
  }

  // There is a real AbortSignal validation later but it doesnt catch falsy primatives.
  if (signal !== true && !signal) {
    throw $ERR_INVALID_ARG_TYPE("signal", "AbortSignal", signal);
  }

  return options;
}

function ReadStream(this: FSStream, path, options): void {
  if (!(this instanceof ReadStream)) {
    return new ReadStream(path, options);
  }
  
  options = copyObject(getStreamOptions(options));

  // Only buffers are supported.
  options.decodeStrings = true;

  let {
    fd, 
    autoClose,
    fs: customFs,
    start = 0,
    end = Infinity,
    encoding,
  } = options;
  if (fd == null) {
    this[kFs] = customFs || fs;
    this.fd = null;
    this.path = getValidatedPath(path);
    const { flags, mode } = options;
    this.flags = flags === undefined ? "r" : flags;
    this.mode = mode === undefined ? 0o666 : mode;
    if (customFs) {
      validateFunction(customFs.open, "options.fs.open");
    }
  } else if (typeof options.fd === "number") {
    // When fd is a raw descriptor, we must keep our fingers crossed
    // that the descriptor won't get closed, or worse, replaced with
    // another one
    // https://github.com/nodejs/node/issues/35862
    if (Object.is(fd, -0)) {
      fd = 0;
    } else {
      validateInt32(fd, "fd", 0, 2147483647);
    }
    this.fd = fd;
    this[kFs] = customFs || fs;
  } else if (typeof fd === 'object' && fd instanceof FileHandle) {
    if (options.fs) {
      throw $ERR_METHOD_NOT_IMPLEMENTED("fs.FileHandle with custom fs operations");
    }
    this[kFs] = fileHandleStreamFs(fd);
    this.fd = fd[kFd];
    fd[kRef]();
    fd.on('close', this.close.bind(this));
  } else {
    throw $ERR_INVALID_ARG_TYPE('options.fd', 'number or FileHandle', fd);
  }

  if (customFs) {
    validateFunction(customFs.read, "options.fs.read");
  }

  $assert(this[kFs], 'fs implementation was not assigned');

  if((options.autoDestroy = autoClose === undefined
    ? true 
    : autoClose) && customFs) {
    validateFunction(customFs.close, "options.fs.close");
  }

  this.start = start;
  this.end = end;
  this.pos = undefined;
  this.bytesRead = 0;

  if (start !== undefined) {
    validateInteger(start, "start", 0);
    this.pos = start;
  }

  if (end === undefined) {
    end = Infinity;
  } else if (end !== Infinity) {
    validateInteger(end, "end", 0);
    if (start !== undefined && start > end) {
      throw $ERR_OUT_OF_RANGE("start", `<= "end" (here: ${end})`, start);
    }
  }
  this[kIsPerformingIO] = false;

  this[kReadStreamFastPath] =
    start === 0 &&
    end === Infinity &&
    autoClose &&
    !customFs &&
    // is it an encoding which we don't need to decode?
    (encoding === "buffer" || encoding === "binary" || encoding == null || encoding === "utf-8" || encoding === "utf8");
  Readable.$call(this, options);
  return this as unknown as void;
}
$toClass(ReadStream, "ReadStream", Readable);
const readStreamPrototype = ReadStream.prototype;

Object.defineProperty(readStreamPrototype, 'autoClose', {
  get() {
    return this._readableState.autoDestroy;
  },
  set(val) {
    this._readableState.autoDestroy = val;
  },
});

const streamNoop = function open() {
  // noop
}
function streamConstruct(this: FSStream, callback: (e?: any) => void) {
  const { fd } = this;
  if (typeof fd === "number") {
    callback();
    return;
  }
  if (this.open !== streamNoop) {
    // Backwards compat for monkey patching open().
    const orgEmit: any = this.emit;
    this.emit = function(...args) {
      if (args[0] === 'open') {
        this.emit = orgEmit;
        callback();
        orgEmit.$apply(this, args);
      } else if (args[0] === 'error') {
        this.emit = orgEmit;
        callback(args[1]);
      } else {
        orgEmit.$apply(this, args);
      }
    } as any;
    this.open();
  } else {
    this[kFs].open(this.path, this.flags, this.mode, (err, fd) => {
      if (err) {
        callback(err);
      } else {
        this.fd = fd;
        callback();
        this.emit('open', this.fd);
        this.emit('ready');
      }
    });
  }
}

readStreamPrototype.open = streamNoop;

readStreamPrototype._construct = streamConstruct;

readStreamPrototype._read = function(n) {
  n = this.pos !== undefined ?
    $min(this.end - this.pos + 1, n) :
    $min(this.end - this.bytesRead + 1, n);

  if (n <= 0) {
    this.push(null);
    return;
  }

  const buf = Buffer.allocUnsafeSlow(n);

  this[kIsPerformingIO] = true;
  this[kFs]
    .read(this.fd, buf, 0, n, this.pos, (er, bytesRead, buf) => {
      this[kIsPerformingIO] = false;

      // Tell ._destroy() that it's safe to close the fd now.
      if (this.destroyed) {
        this.emit(kIoDone, er);
        return;
      }

      if (er) {
        require('internal/streams/destroy').errorOrDestroy(this, er);
      } else if (bytesRead > 0) {
        if (this.pos !== undefined) {
          this.pos += bytesRead;
        }

        this.bytesRead += bytesRead;

        if (bytesRead !== buf.length) {
          // Slow path. Shrink to fit.
          // Copy instead of slice so that we don't retain
          // large backing buffer for small reads.
          const dst = Buffer.allocUnsafeSlow(bytesRead);
          buf.copy(dst, 0, 0, bytesRead);
          buf = dst;
        }

        this.push(buf);
      } else {
        this.push(null);
      }
    });
};

readStreamPrototype._destroy = function(err, cb) {
  // Usually for async IO it is safe to close a file descriptor
  // even when there are pending operations. However, due to platform
  // differences file IO is implemented using synchronous operations
  // running in a thread pool. Therefore, file descriptors are not safe
  // to close while used in a pending read or write operation. Wait for
  // any pending IO (kIsPerformingIO) to complete (kIoDone).
  if (this[kIsPerformingIO]) {
    this.once(kIoDone, (er) => close(this, err || er, cb));
  } else {
    close(this, err, cb);
  }
};

readStreamPrototype.close = function(cb) {
  if (typeof cb === 'function') require('node:stream').finished(this, cb);
  this.destroy();
};

Object.defineProperty(readStreamPrototype, 'pending', {
  get() {
    return this.fd == null;
  },
  configurable: true,
});

function close(stream, err, cb) {
  if (!stream.fd) {
    cb(err);
  } else if (stream.flush) {
    stream[kFs].fsync(stream.fd, (flushErr) => {
      closeAfterSync(stream, err || flushErr, cb);
    });
  } else {
    closeAfterSync(stream, err, cb);
  }
}

function closeAfterSync(stream, err, cb) {
  stream[kFs].close(stream.fd, (er) => {
    cb(er || err);
  });
  stream.fd = null;
}

function WriteStream(this: FSStream, path: string | null, options: any): void {
  if (!(this instanceof WriteStream)) {
    return new WriteStream(path, options);
  }

  options = copyObject(getStreamOptions(options));

  // Only buffers are supported.
  options.decodeStrings = true;

  let {
    fd, 
    autoClose,
    fs: customFs,
    start,
    flush,
  } = options;
  if (fd == null) {
    this[kFs] = customFs || fs;
    this.fd = null;
    this.path = getValidatedPath(path);
    const { flags, mode } = options;
    this.flags = flags === undefined ? "w" : flags;
    this.mode = mode === undefined ? 0o666 : mode;
    if (customFs) {
      validateFunction(customFs.open, "options.fs.open");
    }
  } else if (typeof options.fd === "number") {
    // When fd is a raw descriptor, we must keep our fingers crossed
    // that the descriptor won't get closed, or worse, replaced with
    // another one
    // https://github.com/nodejs/node/issues/35862
    if (Object.is(fd, -0)) {
      fd = 0;
    } else {
      validateInt32(fd, "fd", 0, 2147483647);
    }
    this.fd = fd;
    this[kFs] = customFs || fs;
  } else if (typeof fd === 'object' && fd instanceof FileHandle) {
    if (options.fs) {
      throw $ERR_METHOD_NOT_IMPLEMENTED("fs.FileHandle with custom fs operations");
    }
    this[kFs] = fileHandleStreamFs(fd);
    this.fd = fd[kFd];
    fd[kRef]();
    fd.on('close', this.close.bind(this));
  } else {
    throw $ERR_INVALID_ARG_TYPE('options.fd', 'number or FileHandle', fd);
  }

  const autoDestroy =  options.autoDestroy = autoClose === undefined
    ? true 
    : autoClose;

  if (customFs) {
    const { write, writev, close, fsync } = customFs;
    if (write) validateFunction(write, "options.fs.write");
    if (writev) validateFunction(writev, "options.fs.writev");
    if (autoDestroy) validateFunction(close, "options.fs.close");
    if (flush) validateFunction(fsync, "options.fs.fsync");
    if (!write && !writev) {
      throw $ERR_INVALID_ARG_TYPE("options.fs.write", "function", write);
    }
  } else {
    this._writev = undefined;
    $assert(this[kFs].write, "assuming user does not delete fs.write!");
  }

  if (flush == null) {
    this.flush = false;
  } else {
    if (typeof flush !== "boolean") throw $ERR_INVALID_ARG_TYPE("options.flush", "boolean", flush);
    this.flush = flush;
  }

  this.start = start;
  this.pos = undefined;
  this.bytesWritten = 0;

  if (start !== undefined) {
    validateInteger(start, "start", 0);
    this.pos = start;
  }

  Writable.$call(this, options);

  if (options.encoding) {
    this.setDefaultEncoding(options.encoding);
  }
  return this as unknown as void;
}
$toClass(WriteStream, "WriteStream", Writable);
const writeStreamPrototype = WriteStream.prototype;

writeStreamPrototype.open = streamNoop;
writeStreamPrototype._construct = streamConstruct;

function writeAll(data, size, pos, cb, retries = 0) {
  this[kFs].write(this.fd, data, 0, size, pos, (er, bytesWritten, buffer) => {
    // No data currently available and operation should be retried later.
    if (er?.code === 'EAGAIN') {
      er = null;
      bytesWritten = 0;
    }

    if (this.destroyed || er) {
      return cb(er || $ERR_STREAM_DESTROYED('write'));
    }

    this.bytesWritten += bytesWritten;

    retries = bytesWritten ? 0 : retries + 1;
    size -= bytesWritten;
    pos += bytesWritten;

    // Try writing non-zero number of bytes up to 5 times.
    if (retries > 5) {
      // cb($ERR_SYSTEM_ERROR('write failed'));
      cb(new Error('write failed'));
    } else if (size) {
      writeAll.$call(this, buffer.slice(bytesWritten), size, pos, cb, retries);
    } else {
      cb();
    }
  });
}

function writevAll(chunks, size, pos, cb, retries = 0) {
  this[kFs].writev(this.fd, chunks, this.pos, (er, bytesWritten, buffers) => {
    // No data currently available and operation should be retried later.
    if (er?.code === 'EAGAIN') {
      er = null;
      bytesWritten = 0;
    }

    if (this.destroyed || er) {
      return cb(er || $ERR_STREAM_DESTROYED('writev'));
    }

    this.bytesWritten += bytesWritten;

    retries = bytesWritten ? 0 : retries + 1;
    size -= bytesWritten;
    pos += bytesWritten;

    // Try writing non-zero number of bytes up to 5 times.
    if (retries > 5) {
      // cb($ERR_SYSTEM_ERROR('writev failed'));
      cb(new Error('writev failed'));
    } else if (size) {
      writevAll.$call(this, [Buffer.concat(buffers).slice(bytesWritten)], size, pos, cb, retries);
    } else {
      cb();
    }
  });
}

writeStreamPrototype._write = function(data, encoding, cb) {
  this[kIsPerformingIO] = true;
  writeAll.$call(this, data, data.length, this.pos, (er) => {
    this[kIsPerformingIO] = false;
    if (this.destroyed) {
      // Tell ._destroy() that it's safe to close the fd now.
      cb(er);
      return this.emit(kIoDone, er);
    }

    cb(er);
  });

  if (this.pos !== undefined)
    this.pos += data.length;
};

writeStreamPrototype._writev = function(data, cb) {
  const len = data.length;
  const chunks = new Array(len);
  let size = 0;

  for (let i = 0; i < len; i++) {
    const chunk = data[i].chunk;

    chunks[i] = chunk;
    size += chunk.length;
  }

  this[kIsPerformingIO] = true;
  writevAll.$call(this, chunks, size, this.pos, (er) => {
    this[kIsPerformingIO] = false;
    if (this.destroyed) {
      // Tell ._destroy() that it's safe to close the fd now.
      cb(er);
      return this.emit(kIoDone, er);
    }

    cb(er);
  });

  if (this.pos !== undefined)
    this.pos += size;
};

writeStreamPrototype._destroy = function(err, cb) {
  // Usually for async IO it is safe to close a file descriptor
  // even when there are pending operations. However, due to platform
  // differences file IO is implemented using synchronous operations
  // running in a thread pool. Therefore, file descriptors are not safe
  // to close while used in a pending read or write operation. Wait for
  // any pending IO (kIsPerformingIO) to complete (kIoDone).
  if (this[kIsPerformingIO]) {
    this.once(kIoDone, (er) => close(this, err || er, cb));
  } else {
    close(this, err, cb);
  }
};

writeStreamPrototype.close = function(cb) {
  if (cb) {
    if (this.closed) {
      process.nextTick(cb);
      return;
    }
    this.on('close', cb);
  }

  // If we are not autoClosing, we should call
  // destroy on 'finish'.
  if (!this.autoClose) {
    this.on('finish', this.destroy);
  }

  // We use end() instead of destroy() because of
  // https://github.com/nodejs/node/issues/2006
  this.end();
};

// There is no shutdown() for files.
writeStreamPrototype.destroySoon = writeStreamPrototype.end;

Object.defineProperty(writeStreamPrototype, 'autoClose', {
  get() {
    return this._writableState.autoDestroy;
  },
  set(val) {
    this._writableState.autoDestroy = val;
  },
});

Object.$defineProperty(writeStreamPrototype, 'pending', {
  get() { return this.fd === null; },
  configurable: true,
});

export default { ReadStream, WriteStream, kReadStreamFastPath };