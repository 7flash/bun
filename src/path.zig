const std = @import("std");
const bun = @import("root").bun;

const os = bun.Environment.os;
const strings = bun.strings;

pub fn isRelativePathCrossPlatform(path: []const u8) bool {
    return switch (path.len) {
        0 => false,
        1 => path[0] == '.',
        // "./" is valid on all platforms, and ".\" is valid on Windows only
        // 2 => if (os != .windows)
        //     strings.eqlComptime(path, "./")
        // else
        //     path[0] == '.' and strings.charIsAnySlash(path[1]),
        2 => switch (os) {
            .windows => path[0] == '.' and strings.charIsAnySlash(path[1]),
            else => strings.eqlComptime(path, "./"),
        },
        // same as above but with "../" and "..\"
        else => switch (os) {
            .windows => strings.eqlComptime(path[0..2], "..") and strings.charIsAnySlash(path[2]),
            else => strings.eqlComptime(path, "../"),
        },
    };
}
