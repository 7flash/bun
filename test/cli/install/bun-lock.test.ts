import { spawn, write, file } from "bun";
import { expect, it, beforeAll, beforeEach, afterAll } from "bun:test";
import { access, copyFile, open, writeFile } from "fs/promises";
import { bunExe, bunEnv as env, isWindows, tmpdirSync, VerdaccioRegistry } from "harness";
import { join } from "path";

var verdaccio: VerdaccioRegistry;
var packageDir: string;
var packageJson: string;

beforeAll(async () => {
  verdaccio = new VerdaccioRegistry();
  await verdaccio.start();
});

afterAll(() => {
  verdaccio.stop();
});

beforeEach(async () => {
  ({ packageDir, packageJson } = await verdaccio.createTestDir());
  env.BUN_INSTALL_CACHE_DIR = join(packageDir, ".bun-cache");
  env.BUN_TMPDIR = env.TMPDIR = env.TEMP = join(packageDir, ".bun-tmp");
});

it("should update dependency version literal when no updates are necessary", async () => {
  await Promise.all([
    write(
      packageJson,
      JSON.stringify({
        workspaces: ["packages/*"],
        dependencies: {
          "no-deps": "1.0.0",
        },
      }),
    ),
    write(
      join(packageDir, "packages", "pkg1", "package.json"),
      JSON.stringify({
        name: "pkg1",
        dependencies: {
          "a-dep": "1.0.1",
        },
      }),
    ),
  ]);

  let { exited } = spawn({
    cmd: [bunExe(), "install", "--save-text-lockfile"],
    cwd: packageDir,
    env,
  });

  expect(await exited).toBe(0);

  const firstLockfile = (await file(join(packageDir, "bun.lock")).text()).replaceAll(
    /localhost:\d+/g,
    "localhost:1234",
  );
  expect(firstLockfile).toMatchSnapshot();

  // "no-deps" is updated, but the version still satisfies the resolved
  // package in the lockfile. no install should happen, but the dependency
  // string in the lockfile should be updated.
  await write(
    packageJson,
    JSON.stringify({
      workspaces: ["packages/*"],
      dependencies: {
        "no-deps": "^1.0.0",
      },
    }),
  );

  ({ exited } = spawn({
    cmd: [bunExe(), "install"],
    cwd: packageDir,
    env,
  }));

  expect(await exited).toBe(0);

  expect(await file(join(packageDir, "node_modules", "no-deps", "package.json")).json()).toEqual({
    name: "no-deps",
    version: "1.0.0",
  });

  const secondLockfile = (await file(join(packageDir, "bun.lock")).text()).replaceAll(
    /localhost:\d+/g,
    "localhost:1234",
  );
  expect(firstLockfile).not.toBe(secondLockfile);
  expect(secondLockfile).toMatchSnapshot();

  // now the same with "a-dep" in the workspace
  await write(
    join(packageDir, "packages", "pkg1", "package.json"),
    JSON.stringify({
      name: "pkg1",
      dependencies: {
        "a-dep": "^1.0.1",
      },
    }),
  );

  ({ exited } = spawn({
    cmd: [bunExe(), "install"],
    cwd: packageDir,
    env,
  }));

  expect(await exited).toBe(0);

  expect(await file(join(packageDir, "node_modules", "a-dep", "package.json")).json()).toEqual({
    name: "a-dep",
    version: "1.0.1",
  });

  const thirdLockfile = (await file(join(packageDir, "bun.lock")).text()).replaceAll(
    /localhost:\d+/g,
    "localhost:1234",
  );
  expect(thirdLockfile).not.toBe(secondLockfile);
  expect(thirdLockfile).not.toBe(firstLockfile);
  expect(thirdLockfile).toMatchSnapshot();
});

it("should write plaintext lockfiles", async () => {
  const package_dir = tmpdirSync();

  // copy bar-0.0.2.tgz to package_dir
  await copyFile(join(__dirname, "bar-0.0.2.tgz"), join(package_dir, "bar-0.0.2.tgz"));

  // Create a simple package.json
  await writeFile(
    join(package_dir, "package.json"),
    JSON.stringify({
      name: "test-package",
      version: "1.0.0",
      dependencies: {
        "dummy-package": "file:./bar-0.0.2.tgz",
      },
    }),
  );

  // Run 'bun install' to generate the lockfile
  const installResult = spawn({
    cmd: [bunExe(), "install", "--save-text-lockfile"],
    cwd: package_dir,
    env,
  });
  await installResult.exited;

  // Ensure the lockfile was created
  await access(join(package_dir, "bun.lock"));

  // Assert that the lockfile has the correct permissions
  const file = await open(join(package_dir, "bun.lock"), "r");
  const stat = await file.stat();

  // in unix, 0o644 == 33188
  let mode = 33188;
  // ..but windows is different
  if (isWindows) {
    mode = 33206;
  }
  expect(stat.mode).toBe(mode);

  expect(await file.readFile({ encoding: "utf8" })).toMatchSnapshot();
});

// won't work on windows, " is not a valid character in a filename
it.skipIf(isWindows)("should escape names", async () => {
  const packageDir = tmpdirSync();
  await Promise.all([
    write(
      join(packageDir, "package.json"),
      JSON.stringify({
        name: "quote-in-dependency-name",
        workspaces: ["packages/*"],
      }),
    ),
    write(join(packageDir, "packages", '"', "package.json"), JSON.stringify({ name: '"' })),
    write(
      join(packageDir, "packages", "pkg1", "package.json"),
      JSON.stringify({
        name: "pkg1",
        dependencies: {
          '"': "*",
        },
      }),
    ),
  ]);

  const { exited } = spawn({
    cmd: [bunExe(), "install", "--save-text-lockfile"],
    cwd: packageDir,
    stdout: "ignore",
    stderr: "ignore",
    env,
  });

  expect(await exited).toBe(0);

  expect(await file(join(packageDir, "bun.lock")).text()).toMatchSnapshot();
});
