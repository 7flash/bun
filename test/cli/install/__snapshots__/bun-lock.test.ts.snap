// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should update dependency version literal when no updates are necessary 1`] = `
"{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "dependencies": {
        "no-deps": "1.0.0",
      },
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "a-dep": "1.0.1",
      },
    },
  },
  "packages": {
    "a-dep": ["a-dep@1.0.1", "http://localhost:1234/a-dep/-/a-dep-1.0.1.tgz", {}, "sha512-6nmTaPgO2U/uOODqOhbjbnaB4xHuZ+UB7AjKUA3g2dT4WRWeNxgp0dC8Db4swXSnO5/uLLUdFmUJKINNBO/3wg=="],

    "no-deps": ["no-deps@1.0.0", "http://localhost:1234/no-deps/-/no-deps-1.0.0.tgz", {}, "sha512-v4w12JRjUGvfHDUP8vFDwu0gUWu04j0cv9hLb1Abf9VdaXu4XcrddYFTMVBVvmldKViGWH7jrb6xPJRF0wq6gw=="],

    "pkg1": ["pkg1@workspace:packages/pkg1", { "dependencies": { "a-dep": "1.0.1" } }],
  }
}
"
`;

exports[`should update dependency version literal when no updates are necessary 2`] = `
"{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "dependencies": {
        "no-deps": "^1.0.0",
      },
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "a-dep": "1.0.1",
      },
    },
  },
  "packages": {
    "a-dep": ["a-dep@1.0.1", "http://localhost:1234/a-dep/-/a-dep-1.0.1.tgz", {}, "sha512-6nmTaPgO2U/uOODqOhbjbnaB4xHuZ+UB7AjKUA3g2dT4WRWeNxgp0dC8Db4swXSnO5/uLLUdFmUJKINNBO/3wg=="],

    "no-deps": ["no-deps@1.0.0", "http://localhost:1234/no-deps/-/no-deps-1.0.0.tgz", {}, "sha512-v4w12JRjUGvfHDUP8vFDwu0gUWu04j0cv9hLb1Abf9VdaXu4XcrddYFTMVBVvmldKViGWH7jrb6xPJRF0wq6gw=="],

    "pkg1": ["pkg1@workspace:packages/pkg1", { "dependencies": { "a-dep": "1.0.1" } }],
  }
}
"
`;

exports[`should update dependency version literal when no updates are necessary 3`] = `
"{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "dependencies": {
        "no-deps": "^1.0.0",
      },
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "a-dep": "^1.0.1",
      },
    },
  },
  "packages": {
    "a-dep": ["a-dep@1.0.1", "http://localhost:1234/a-dep/-/a-dep-1.0.1.tgz", {}, "sha512-6nmTaPgO2U/uOODqOhbjbnaB4xHuZ+UB7AjKUA3g2dT4WRWeNxgp0dC8Db4swXSnO5/uLLUdFmUJKINNBO/3wg=="],

    "no-deps": ["no-deps@1.0.0", "http://localhost:1234/no-deps/-/no-deps-1.0.0.tgz", {}, "sha512-v4w12JRjUGvfHDUP8vFDwu0gUWu04j0cv9hLb1Abf9VdaXu4XcrddYFTMVBVvmldKViGWH7jrb6xPJRF0wq6gw=="],

    "pkg1": ["pkg1@workspace:packages/pkg1", { "dependencies": { "a-dep": "^1.0.1" } }],
  }
}
"
`;

exports[`should write plaintext lockfiles 1`] = `
"{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "test-package",
      "dependencies": {
        "dummy-package": "file:./bar-0.0.2.tgz",
      },
    },
  },
  "packages": {
    "dummy-package": ["bar@./bar-0.0.2.tgz", {}],
  }
}
"
`;

exports[`should escape names 1`] = `
"{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "quote-in-dependency-name",
    },
    "packages/\\"": {
      "name": "\\"",
    },
    "packages/pkg1": {
      "name": "pkg1",
      "dependencies": {
        "\\"": "*",
      },
    },
  },
  "packages": {
    "\\"": ["\\"@workspace:packages/\\"", {}],

    "pkg1": ["pkg1@workspace:packages/pkg1", { "dependencies": { "\\"": "*" } }],
  }
}
"
`;
